Retrieve each commit only once in the commit extraction

bockthom opened the pull request:
As Codeface analyzes the data in three-month ranges, commits which are
on the border of those ranges are included twice as they belong to two
different ranges and, therefore, end up with two different commit ids in
the database though having the same commit hash.

(It is not enough to only filter those commits based on the commit
hashes describing the start or end of a Codeface range. The reason for
that is that Codeface assigns commits to a range not based on the commit
hash describing the start/end of a range, but on the commit date (not
the author date) of the belonging to the commit hash which desribes the
start/end. Hence, all those commits which have the same commit date or
also assigned to both ranges!)

To mitigate this, only extract such commits once, that is, just take the
entries having the lower commit id of those commits having the same
hashes.

(It is not necessary to adjust the range extraction as within each
range, each commit appears already only once. However, keep in mind that
when combining all the range extractions, the commits at the border will
still be duplicated.)

This fixes #28.

bockthom committed:
Retrieve each commit only once in the commit extraction

As Codeface analyses the data in three-month ranges, commits which are
on the border of those ranges are included twice as they belong to two
different ranges and, therefore, end up with two different commit ids in
the database though having the same commit hash.

(It is not enough to only filter those commits based on the commit
hashes describing the start or end of a Codeface range. The reason for
that is that Codeface assigns commits to a range not based on the commit
hash describing the start/end of a range, but on the commit date (not
the author date) of the belonging to the commit hash which desribes the
start/end. Hence, all those commits which have the same commit date or
also assigned to both ranges!)

To mitigate this, only extract such commits once, that is, just take the
entries having the lower commit id of those commits having the same
hashes.

(It is not necessary to adjust the range extraction as within each
range, each commit appears already only once. However, keep in mind that
when combining all the range extractions, the commits at the border will
still be duplicated.)

This fixes #28.

Signed-off-by: Thomas Bock <bockthom@cs.uni-saarland.de>

clhunsen commented:
Nice work, @bockthom! Clean and simple solution. :+1: 

Nevertheless, I would like to discuss one thing: 
> (It is not necessary to adjust the range extraction as within each
> range, each commit appears already only once. However, keep in mind that
> when combining all the range extractions, the commits at the border will
> still be duplicated.)

I am not sure whether this is a good idea. Until now, if I remember correctly, we aimed at keeping both project-level and range-level extractions "in sync", such that you could combine the range-level extractions to construct the project-level data (at least, to a certain extent). If we followed this principle, we would need to remove the duplicated commits here too. (Unfortunately, a solution for this is much harder to implement, I guess.)

Is there a specific reason why you did not implement this for the range-level extraction (except for the needed complicated implementation)? 

----------------------

I do not think that it is _mandatory_ to implement this, but we should, at least, discuss it. :wink:

bockthom commented:
> Is there a specific reason why you did not implement this for the range-level extraction (except for the needed complicated implementation)?

No, there is no specific reason. I just was too lazy to implement this.

I thought about this again, and now I come to the conclusion that we should fix this also for the range-level extraction. Yesterday I've thought that this is not necessary, but after your comment I now think that it should be fixed for the range-level extraction, just for consistency reasons. Thanks for your comment (and, therefore, thanks for criticizing my laziness -- you have convinced me now that it is worth fixing it).

I already have thought about a solution for the range-level extraction in the last minutes. I also try to fix this within the corresponding SQL statement, I already have some ideas how this could be achieved. Let's see if I get that implemented in such a way. I will probably work on that tomorrow.

clhunsen commented:
Have you thought about a sub-statement in the SQL statement to select commit hashes from the previous range? This way, we could select only commit information where the commit hash is _not_ in the previous range. Maybe, the `where` clause in the sub-statement needs to select on the `release_timeline` table to assess the previous range's data. :thinking:

bockthom committed:
Retrieve each commit only once in the range-level commit extraction

This is a follow-up commit for 80e08d0596a14a2fea6e568253e9bb421ae21c95
which makes sure that also for range-level extractions, commits are only
extracted once.

Signed-off-by: Thomas Bock <bockthom@cs.uni-saarland.de>

bockthom committed:
Remove duplicated commits also from commit-message extractions and implementation extractions

Signed-off-by: Thomas Bock <bockthom@cs.uni-saarland.de>

bockthom committed:
Prevent e-mails to be part of two ranges in range-level e-mail extraction

Signed-off-by: Thomas Bock <bockthom@cs.uni-saarland.de>

bockthom commented:
Meanwhile, I have found a solution for the range-level extraction. It was not that easy to get the previous release range, but I have found a solution. The solution works under the assumption that the previous range ends exactly at the same time as the current range begins. For usual three-month ranges, that holds. I already tested it and the duplicated commits get removed from the second range the appear in.

While working on that I noticed that we also need to adjust the commit message extraction and the implementation extraction, as both of those also rely on the commit data and commit ids, which can be duplicated. So, I added a similar fix to both extractions (and both project-level and range-level extractions).

In addition, I noticed that for the e-mail range-level extraction we could have a similar problem: The `BETWEEN` operator in SQL is inclusive, that is, e-mail whose date is exactly the date of such a range border would also be part of both ranges. Here, I also added a similar fix than for the commit extractions.

So, everything should be fixed now, I guess (except for the range-level extraction of authors, which still is not possible on a range-level but just returns everything, as before).

Could you please have a quick look at my fixes, @clhunsen? Thanks a lot in advance! Hopefully, this pull-request is fine now.

bockthom committed:
Remove superfluous join from range level extractions

As we do not need the release-timeline data of the previous range in
range-level extractions on commits, we can remove the corresponding JOIN
in the SQL statements.

Thanks to @clhunsen for pointing this out.

In addition, add a short comment on range bounds: Ranges include their
ends as Codeface uses the committer date of a commit to split the ranges.
Commits at the end may, therefore, have an earlier author date, which is
the reason that such commits better suit into the previous range than the
subsequent range.

Signed-off-by: Thomas Bock <bockthom@cs.uni-saarland.de>

bockthom commented:
As @clhunsen personally told me that he is fine with this PR as soon as the superfluous joins were removed (which is done in commit 939ac9402aeae4331f82c42f7d9900c5a1077f43), I will merge now.
